### ISSUE TITLE
qgmailnotifier collapsed

### ISSUE THREAD

**Issue by p1839251** [eteran/qgmailnotifier#8]
Hello,

It's about time qgmailnotifier was updated. It's not reliable on the latest Plasma desktop any more.
Right click menu totally chaotic. 

Unread Conversation shows irrelevant information often outside of the screen. Basically old mails. There are multiple Abouts and Unread conversations on the list. 
This has been an issue for the last 2 or 3 months.

I use Arch. (qt5-base 5.8.0-7)
Qgmailnotifier has been installed from AUR: https://aur.archlinux.org/packages/qgmailnotifier-git/

**Comment by eteran** [comment]
So the main issue that this isn't a tool I use myself anymore. I am happy to try to improve qgmailnotifier but first I must try to replicate the issues.

In general, please report things as you observe them (and one issue per report so it's easier to track) so that they don't go un-fixed indefinitely.

**Comment by p1839251** [comment]
I felt you hadn't used it any more, since these issues are hard not to be noticed.
I think they are pretty much one problem but I don't know for sure.
A few months ago, everything worked great.

Running from Konsole (terminal), Qgmailnotifier doesn't say anything.
It sometimes loses the ability to show unread conversations.
It occasionally renders unread conversations outside of the screen. The system tray is on the right side of the screen and it tries to display the conversations on the right side too.
Very frequently there are more exits, unread conversations etc on the list than it should be.

To sum up, I think the menu has some problems.
I don't know what triggers the malfunctions. The menu can change without reboot the machine.
How could I be more help? Should I take some screenshots?
Thank you for all your effort!

**Comment by eteran** [comment]
Some screenshots would definitely be helpful. Also, are you on a single or multi monitor setup? If multi, is the panel on the right or left screen (or other)?

Just trying to imagine what could have changed to break things.

**Comment by p1839251** [comment]
Ok, here we go.
To be perfectly frank, I have no idea what's going on, so keep asking questions if you have the slightest idea.
So the problem is; qgmailnotifier is rather inconsistent.
Single monitor setup, there's no other monitor here.
Qgmailnotifier worked fine a few months ago.
I could reproduce all these symptoms on a virtual machine.

First, it wouldn't show the unread conversations. There were some, though.
https://ptpb.pw/7uCQ

In this picture it is working quite well
https://ptpb.pw/Wp2t

shortly afterward, chaos thrived
https://ptpb.pw/Pcqv
Two exits, and shows wrong information about the number of messages. I have 2 new messages, it's visible here:
https://ptpb.pw/L6bw
Since "Me : Music" is rather short it's not visible but if it were longer, it would still stay the right side of the menu bar, therefore it would be outside of the screen.

It sometimes miscalculating the size of the notification
https://ptpb.pw/qP2w
But this is really minor bug. However, this information is usually accurate.

**Comment by eteran** [comment]
Hmm, is it possible that you've recently updated Qt? Because you had said that things worked fine until recently. Is it possible that you have a buggy version of Qt somehow? Stranger things have happened to me :-/.

I ask because looking at some of the code, I am having a hard time seeing where it could be going wrong. For example, this is the code to construct the menu:

	QMenu *const menu = new QMenu(this);
	QAction *const view  = menu->addAction(tr("View Inbox"), this, SLOT(doView()));
	menu->addAction(tr("Compose a Message"), this, SLOT(doCompose()));
	menu->addAction(tr("Check Mail Now"), this, SLOT(doCheck()));
	menu->addAction(tr("Tell me Again..."), this, SLOT(doTell()));
	menu->addAction(tr("Options"), this, SLOT(doOptions()));
	menu->addAction(tr("About"), this, SLOT(doAbout()));

	if(currentMails_.size() != 0) {

		menu->addSeparator();
		QMenu *const convsations = menu->addMenu(tr("Unread Conversations"));

		QSettings settings;

		const int maxCons = settings.value("max_conversations", 10).value<int>();

		int index = 0;
		Q_FOREACH(GMailEntry entry, currentMails_) {
			QAction *const action = convsations->addAction(entry.author_name + " : " + entry.title, this, SLOT(readConversation()));
			action->setData(index++);
			if(index >= maxCons) {
				break;
			}
		}
	}

	menu->addSeparator();
	menu->addAction(tr("Exit"), qApp, SLOT(quit()));

	menu->setDefaultAction(view);
	return menu;

I don't know if you are much of a coder, but the main take away is there is one line which adds the "Exit" action, and it isn't in a loop and it isn't conditional. So, how could it possibly happen twice?

Likewise, the actual notifications are done by the Qt function `QSystemTrayIcon::showMessage(...)`, which is a function Qt itself provides. So, while it's possible that somehow I'm misusing it, it is not obvious to me how that would be :-/

I'll keep digging for a bit, I'm having a hard time reproducing the issue and generally diagnosing the cause of the problem from a code point of view.

**Comment by p1839251** [comment]
It's quite likely that I updated Qt. Archlinux is a rolling release distro, every package will be updated sooner or later. I'm on stable branch btw. Although I don't know what Qt package you referred to. I mean, does qgmailnotifier depend on Qt4 or Qt5? I updated both on 7 April. Qt4 4.8.7-12 to 4.8.7.16 and Qt5 5.8.0-5 to 5.8.0-7. Neither seems to be a major update.
I'm not much of a coder. Little do I know about coding. As I saw the code I had some idea what was going on but I can't give you any advice.
I'm fairly certain that an update messed qgmailnotifier up. If you can't reproduce this, you might use older packages. I've just checked qgmailnotifier now. https://ptpb.pw/eh99 Exit has just gone away.

I tried to downgrade Qt5 back to 5.8.0-5 but then I wasn't able to run qgmailnotifier any more.
qgmailnotifier: error while loading shared libraries: libicui18n.so.58: cannot open shared object file: No such file or directory. This issue seems to be rather complex.

**Comment by p1839251** [comment]
It's very fascinating now.
https://ptpb.pw/Bu3j

'Unread Conversations' keeps telling me outdated information even when I refresh mails (check mail now). Mostly read and deleted messages.
Hmm. Isn't there a syntax error or a missing quote or I don't know around there?

To summarize, view inbox, compose message, check mail, tell me again, options and about all work.
Unread Conversations almost never works. Exit(s) are sometimes naughty.
I suspect something is wrong with unread conversations.

Other Qt applications work fine.

**Comment by eteran** [comment]
Ok, so I think I'll need to just create a VM of the same version of arch Linux as you and see if I can reproduce it that way.

So, please let me know as specific as possible which version you are using, and whether it is 32 or 64 bit.

Hopefully then, I can see in action the aymtoms you are seeing.

**Comment by p1839251** [comment]
Since installing Archlinux is not easy and rather time consuming I recommend you install an easier alternative such as Antergos which is basically Arch and you can have a similar environment then.
I'm kind of surprised that you don't have similar problems.

There aren't really Archlinux versions, it's a rolling release distro. It's 64 bit.
Here's a list about my Qt packages. You can see the versions here:
https://ptpb.pw/ut85
It would be really difficult to obtain the same packages. I think an updated Arch is enough.
I use KDE Plasma 5 environment.

Here's the PKGBUILD that I used for installing qgmailnotifier:
https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=qgmailnotifier-git

**Comment by eteran** [comment]
I'm a gentoo guy, so rolling release is no problem in my mind :-)

I'll setup a VM sometime this weekend and see what I can do.

**Comment by p1839251** [comment]
I provide a VBox file for you. Qmailnotifier has been already installed.
https://www.dropbox.com/s/efa4rrq40b6mjcc/qgmail.tar.bz2?dl=0

Issue: https://ptpb.pw/y7D0
It sometimes pretends that everything is okay, you need to play with it.
However, the arrow '>' is often missing.

**Comment by eteran** [comment]
Thanks! I downloaded the image. I'm a bit busy with work this week, so I may not get to take a look right away. This will be VERY helpful in reproducing the issue though!

**Comment by p1839251** [comment]
It's weird but I hope you will also have problems with it because I managed to use this virtual machine for hours without any problems. I really don't know what cause the malfunctions.
That's the only app currently that has such problems so I can't say it's a Qt or even KDE bug.

For the time being, qgmailnotifier is working well on my (host) machine. I haven't updated my system since then. I can't work out what triggers this effect.
The 'mouse over notification' is often disproportionate though. Nevertheless is so minor that I wouldn't have filed a bug report because of this. 

!!As a matter of fact I've just found something in this very moment!! I said there was no problems, didn't I? Well, it's isn't true any more.

**So the steps to reproduce this**:
Qgmailnotifier was running - obviously.
I had one unread e-mail. Both mouse-over-notification and unread-conversations> claimed the same.
I sent myself an e-mail and I didn't select 'check the mailbox now' but I assume the latter is optional. I was waiting to be checked automatically. I set the check interval to 600000.
As soon as the new mail arrived Qgmailnotifier has gone crazy.
I have 2 exits now and according to Unread Conversations I have only 1 unread messages.
(Mouse over notification shows the right number, btw)
